File: src/main/java/com/github/androidutils/wakelock/WakeLockManager.java
/*
 * Copyright (C) 2012 Yuriy Kulikov yuriy.kulikov.87@gmail.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.androidutils.wakelock;

import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.CopyOnWriteArrayList;

import android.app.Application;
import android.content.Context;
import android.content.Intent;
import android.os.PowerManager;
import android.os.PowerManager.WakeLock;

import com.github.androidutils.logger.Logger;
import com.google.inject.Inject;
import com.google.inject.name.Named;

/**
 * Utility class to pass {@link WakeLock} objects with intents. It contains a
 * factory, which has to be initialized in {@link Application#onCreate()} or
 * otherwise there will be an exception. Instance maintains a {@link Map} of
 * String tag to {@link WakeLock} wakelock
 */
public class WakeLockManager {
    public static final String EXTRA_WAKELOCK_TAG = "WakeLockManager.EXTRA_WAKELOCK_TAG";
    public static final String EXTRA_WAKELOCK_HASH = "WakeLockManager.EXTRA_WAKELOCK_HASH";

    private final Logger log;
    private final CopyOnWriteArrayList<WakeLock> wakeLocks;
    private final PowerManager pm;

    @Inject
    public WakeLockManager(Context context, @Named("debug") Logger logger) {
        pm = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
        wakeLocks = new CopyOnWriteArrayList<PowerManager.WakeLock>();
        log = logger;
    }

    /**
     * Acquires a partial {@link WakeLock}, stores it internally and puts the
     * tag into the {@link Intent}. To be used with
     * {@link WakeLockManager#releasePartialWakeLock(Intent)}
     * 
     * @param intent
     * @param wlTag
     */
    public void acquirePartialWakeLock(Intent intent, String wlTag) {
        final WakeLock wl = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, wlTag);
        wl.acquire();
        wakeLocks.add(wl);
        intent.putExtra(WakeLockManager.EXTRA_WAKELOCK_HASH, wl.hashCode());
        intent.putExtra(WakeLockManager.EXTRA_WAKELOCK_TAG, wlTag);
        log.d(wl.toString() + " " + wlTag + " was acquired");
    }

    /**
     * Releases a partial {@link WakeLock} with a tag contained in the given
     * {@link Intent}
     * 
     * @param intent
     */
    public void releasePartialWakeLock(Intent intent) {
        if (intent.hasExtra(WakeLockManager.EXTRA_WAKELOCK_TAG)) {
            final int hash = intent.getIntExtra(WakeLockManager.EXTRA_WAKELOCK_HASH, -1);
            final String tag = intent.getStringExtra(WakeLockManager.EXTRA_WAKELOCK_TAG);
            // We use copy on write list. Iterator won't cause
            // ConcurrentModificationException
            for (Iterator<WakeLock> iterator = wakeLocks.iterator(); iterator.hasNext();) {
                WakeLock wakeLock = iterator.next();
                if (hash == wakeLock.hashCode()) {
                    if (wakeLock.isHeld()) {
                        wakeLock.release();
                        log.d(wakeLock.toString() + " " + tag + " was released");
                    } else {
                        log.e(wakeLock.toString() + " " + tag + " was already released!");
                    }
                    wakeLocks.remove(wakeLock);
                    return;
                }
            }
            log.e("Hash " + hash + " was not found");
        }
    }
}

File: app/src/main/java/com/better/alarm/platform/WakeLockManager.kt
/*
 * Copyright (C) 2012 Yuriy Kulikov yuriy.kulikov.87@gmail.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.better.alarm.platform

import android.app.Application
import android.content.Intent
import android.os.PowerManager
import android.os.PowerManager.WakeLock
import com.better.alarm.logger.Logger
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.atomic.AtomicInteger

/**
 * Utility class to pass [WakeLock] objects with intents. It contains a factory, which has to be
 * initialized in [Application.onCreate] or otherwise there will be an exception. Instance maintains
 * a [Map] of String tag to [WakeLock] wakelock
 */
class WakeLockManager(private val log: Logger, val pm: PowerManager) : Wakelocks {
  private val wakelockCounter = AtomicInteger(0)
  private val wakeLockIds = CopyOnWriteArrayList<Int>()
  private val serviceWakelock =
      pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "SimpleAlarmClock:AlertServiceWrapper")
  private val transitionWakelock =
      pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "SimpleAlarmClock:AlertServicePusher")

  override fun acquireServiceLock() {
    log.debug { "Acquired service wakelock" }
    serviceWakelock.acquire(60 * 60_000)
  }

  override fun releaseServiceLock() {
    if (serviceWakelock.isHeld) {
      log.debug { "Released service wakelock" }
      serviceWakelock.release()
    }
  }

  /**
   * Acquires a partial [WakeLock], stores it internally and puts the tag into the [Intent]. To be
   * used with [WakeLockManager.releaseTransitionWakeLock]
   */
  fun acquireTransitionWakeLock(intent: Intent) {
    transitionWakelock.acquire(60 * 1000)
    wakelockCounter.incrementAndGet().also { count ->
      wakeLockIds.add(count)
      intent.putExtra(COUNT, count)
      log.debug { "Acquired $transitionWakelock #$count" }
    }
  }

  /**
   * Releases a partial [WakeLock] with a tag contained in the given [Intent]
   *
   * @param intent
   */
  fun releaseTransitionWakeLock(intent: Intent) {
    val count = intent.getIntExtra(COUNT, -1)
    val wasRemoved = wakeLockIds.remove(count)
    if (wasRemoved && transitionWakelock.isHeld) {
      transitionWakelock.release()
      log.debug { "Released $transitionWakelock #$count" }
    }
  }

  companion object {
    const val COUNT = "WakeLockManager.COUNT"
  }
}
